tools/pipeline-witness/infra/kusto/views/Metrics.kql

```diff
@@ -0,0 +1,14 @@
+// Create materialized view
+.create materialized-view Metrics on table BuildLogLine {
+BuildLogLine
+    | where Message has_all ("logmetric", "name", "value", "timestamp")
+    | where Message startswith "logmetric:"
+    | extend json = todynamic(extract(@"^(?i:logmetric):\s*(.*?)\s*$", 1, Message))
+    | where not(isnull(json["name"]))
+    | extend
+        Name = tostring(json.name),
+        Labels = todynamic(json.labels),
+        Value = toreal(json.value),
+        Timestamp = todatetime(json.timestamp)
+    | summarize arg_max(LogId, LineNumber, Name, Value, Timestamp, Labels) by OrganizationName, ProjectId, BuildId, Hash=hash_sha1(Message)
```

Comment:
I'm curious as to why `arg_max` is used here? Are we expecting duplicate metrics might be published?

Reply:
It's in the Prometheus and InfluxDB specs.  Metrics with identical keys, value, and time get deduped.  We do see duplicate log line ingestion (Task vs Job), so it's good to dedupe.

Reply:
I see, so because you are doing arg_max by BuildId we should expect LogId to be a useful key here.

Reply:
It's really by BuildId/Hash(message) that matters most.     If any build produces 2 identical metric statement (same name, value, timestamp and tags) we dedupe and favor the one with the highest log id